modules/table/src/main/java/org/apache/ignite/internal/table/distributed/LowWatermarkImpl.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.schema.configuration.LowWatermarkConfiguration;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.tx.TxManager;
+import org.apache.ignite.internal.util.ByteUtils;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.internal.vault.VaultEntry;
+import org.apache.ignite.internal.vault.VaultManager;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Class to manage the low watermark.
+ *
+ * <p>Low watermark is the node's local time, which ensures that read-only transactions have completed by this time, and new read-only
+ * transactions will only be created after this time, and we can safely delete obsolete/garbage data such as: obsolete versions of table
+ * rows, remote indexes, remote tables, etc.
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/IGNITE/IEP-91%3A+Transaction+protocol">IEP-91</a>
+ */
+public class LowWatermarkImpl implements IgniteComponent, LowWatermark {
+    private static final IgniteLogger LOG = Loggers.forClass(LowWatermarkImpl.class);
+
+    static final ByteArray LOW_WATERMARK_VAULT_KEY = new ByteArray("low-watermark");
+
+    private final LowWatermarkConfiguration lowWatermarkConfig;
+
+    private final HybridClock clock;
+
+    private final TxManager txManager;
+
+    private final VaultManager vaultManager;
+
+    private final List<LowWatermarkChangedListener> updateListeners = new CopyOnWriteArrayList<>();
+
+    private final ScheduledExecutorService scheduledThreadPool;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    private volatile @Nullable HybridTimestamp lowWatermark;
+
+    private final AtomicReference<ScheduledFuture<?>> lastScheduledTaskFuture = new AtomicReference<>();
+
+    private final FailureProcessor failureProcessor;
+
+    /**
+     * Constructor.
+     *
+     * @param nodeName Node name.
+     * @param lowWatermarkConfig Low watermark configuration.
+     * @param clock A hybrid logical clock.
+     * @param txManager Transaction manager.
+     * @param vaultManager Vault manager.
+     * @param failureProcessor Failure processor tha is used to handle critical errors.
+     */
+    public LowWatermarkImpl(
+            String nodeName,
+            LowWatermarkConfiguration lowWatermarkConfig,
+            HybridClock clock,
+            TxManager txManager,
+            VaultManager vaultManager,
+            FailureProcessor failureProcessor
+    ) {
+        this.lowWatermarkConfig = lowWatermarkConfig;
+        this.clock = clock;
+        this.txManager = txManager;
+        this.vaultManager = vaultManager;
+        this.failureProcessor = failureProcessor;
+
+        scheduledThreadPool = Executors.newSingleThreadScheduledExecutor(
```

Comment:
It would be necessary to create a task with todo to use a common pool and not a separate one.

Reply:
The class was renamed, the content was mostly left as is.

Reply:
Please create a ticket.

Reply:
https://issues.apache.org/jira/browse/IGNITE-21744